src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupBasedAlleles.java

```diff
@@ -31,47 +54,132 @@
+                }
             }
         }
 
         return pileupSNPsList;
     }
 
+    /**
+     * Apply the filters to discovered alleles
+     * - Does it have greater than snpThreshold percentage of bases support in the pileups?
+     * - Does it have greater than pileupAbsoluteDepth number of reads supporting it?
+     * - Are the reads supporting alts at the site greater than badReadThreshold percent "good"? //TODO evaluate if this is worth doing on a per-allele basis or otherwise
+     */
+    private static boolean passesFilters(final PileupDetectionArgumentCollection args, boolean indel,  final int numOfBases, final int totalAltBadReads, final int totalAltReads, final Map.Entry<?, Integer> maxAlt) {
+        return ((float) maxAlt.getValue() / (float) numOfBases) > (indel ? args.indelThreshold : args.snpThreshold)
+                && numOfBases >= args.pileupAbsoluteDepth
+                && ((args.badReadThreshold <= 0.0) || (float) totalAltBadReads / (float)totalAltReads <= args.badReadThreshold);
+    }
+
+    /**
+     * Based on the illumina PileupDetection filtering code: We apply a number of configurable heuristics to the reads that support
+     * alt alleles that may be added and for each read evaluate if its "bad" or not by each of the heurisitcs. Currently they are:
+     * - Secondary/SA tag reads are bad
+     * - Improperly paired reads are bad
+     * - Reads with > 8% per-base edit distance to the reference are bad
+     * - Reads 2 std deviations away from the standard insert size are bad (not implemented)
+     *
+     * @param read
+     * @param referenceContext
+     * @param args
+     * @param headerForRead TODO get rid of this sam record coversion
+     * @return
+     */
+    @VisibleForTesting
+    static boolean evaluateBadRead(final GATKRead read, final ReferenceContext referenceContext, final PileupDetectionArgumentCollection args, final SAMFileHeader headerForRead) {
+        if (args.badReadProperPair && !read.isProperlyPaired()) {
+            return true;
+        }
+        if (args.badReadSecondary && (read.isSecondaryAlignment() || read.hasAttribute("SA"))) {
+            return true;
+        }
+
+
+        //TODO this conversion is really unnecessary...
+        SAMRecord samRecordForRead = read.convertToSAMRecord(headerForRead);
+
+        // Assert that the edit distance for the read is in line
+        if (args.badReadEditDistance > 0.0) {
+            if (! read.hasAttribute("NM")) {
+                SequenceUtil.calculateSamNmTag(samRecordForRead, referenceContext.getBases(new SimpleInterval(read)));
+            }
+            int mismatches = read.getAttributeAsInteger("NM");
+            if (mismatches > (read.getLength() * args.badReadEditDistance)) {
+                return true;
+            }
+        }
+
+        //TODO add threshold descibed by illumina about insert size compared to the average
+        if (args.templateLenghtStd > 0 && args.templateLengthMean > 0) {
```

Comment:
insert-size is a much better term than template length.....i think we should use that instead.

Reply:
We are already using `samRecordForRead.getInferredInsertSize()` or do you mean we should change the title/arguments for this to use "insert size" as a name?